File: m/rls/clarg.r
; ???	bal is overwritten
; ???	sep is not used
;Wnt := 1
code	cl_arg - returns nth argument 
include	rid:rider
include	rid:cldef
include	rid:ctdef
include	rid:medef
include	rid:mxdef
include	rid:stdef

	cl_ptr : (int, *char, int, *char, **char) int-

code	cl_arg - get nth argument from string

  func	cl_arg
	nth : int			; the argument required
	str : * char			; string to vectorize (and destroy)
	dst : * char			; result pointer
	sep : int			; the separator
	bal : * char			; the balance set
	()  : int			;
  is	src : * char			;
	len : int
	*dst = 0			; assume nothing
	len = cl_ptr(nth,str,sep,bal,&src); decode it
	pass fail			; nothing doing
	*<*char>me_mov (src,dst,len) = 0; copy it
	fine				;
  end

code	cl_ptr - get pointer to nth argument

  func	cl_ptr
	nth : int			; the argument required
	str : * char~			; string to vectorize (and destroy)
	sep : int			;
	bal : * char~			;
	res : ** char			; result pointer
	()  : int own			; length or fail
  is	fail if nth lt			; negative index
	sep = ' ' if sep eq		; default separator
	bal = "\"\'" if bal eq		;
      repeat				; the nth
	++str while ct_spc (*str)	; skip leading spaces
	quit if *str eq			; all done
	*res = str			; save start address
	while *str ne			; got more
	   bal = "\"\'"		;???	; balance set
	   while *bal ne		; look for balance member
	      if *str eq *bal++		; balance this one?
	   .. .. quit str = st_bal (str);
	   quit if *str eq ' ' || !*str	; gotcha or eol (from st_bal)
	   ++str			; try again
	end				;
 	quit if nth-- eq		; found it
	++str if *str ne		; skip terminator
      forever				; look at next
	fail if nth ge			; failed to find our dear one
	reply str-*res			; return the length of it
  end
end file
code	cl_vec - convert to string vector

  func	cl_vec
	str : * char			;
	cnt : * int			; result count
	ptr : *** char			; result string vector
	()  : int			; fine/fail
  is	nth : int = 0			;
	vec : ** char = *ptr		;
	if vec eq <>			; not allocated yet
	   vec = me_alc (33 * #<*char>)	; upto 32 parameters	   
	.. *ptr = vec			;
					;
	*cnt = 0			;
	*vec = st_dup ("")		; null string for first
	repeat				;
	   *++vec = <>			; terminate list
	   fail if *++cnt eq 32		; count and limit it
	   cl_nth (nth, str, *vec)	; get the nth 
	   fine if fail			; that was the last
	end				;
	fine				;
  end

code	cl_arg - get nth argument from string

  func	cl_arg
	nth : int			; the argument required
	str : * char			; string to vectorize (and destroy)
	dst : * char			; result pointer
	sep : int			; the separator
	bal : * char			; the balance set
	()  : int			;
  is	src : * char			;
	len : int
	len = cl_ptr(nth,str,sep,bal,&src); decode it
	pass fail			; nothing doing
	*<*char>me_mov (src,dst,len) = 0; copy it
	fine				;
  end

code	cl_ptr - get pointer to nth argument

  func	cl_ptr
	nth : int			; the argument required
	str : * char			; string to vectorize (and destroy)
	sep : int			;
	bal : * char			;
	res : ** char			; result pointer
	()  : int own			; length or fail
  is	fail if nth lt			; negative index
	sep = ' ' if sep eq		; default separator
	bal = "\"\'" if bal eq		;
      repeat				; the nth
	++str while ct_spc (*str)	; skip leading spaces
	quit if *str eq			; all done
	*res = str			; save start address
	while *str ne			; got more
	   bal = "\"\'"		;???	; balance set
	   while *bal ne		; look for balance member
	      if *str eq *bal++		; balance this one?
	   .. .. quit str = st_bal (str);
	   quit if *str eq ' ' || !*str	; gotcha or eol (from st_bal)
	   ++str			; try again
	end				;
 	quit if nth-- eq		; found it
	++str if *str ne		; skip terminator
      forever				; look at next
	fail if nth			; failed to find our dear one
	reply str-*res			; return the length of it
  end


File: m/rls/rlb/clarg.c
/* code -  cl_arg - returns nth argument */
#include "m:\rid\rider.h"
#include "m:\rid\cldef.h"
#include "m:\rid\ctdef.h"
#include "m:\rid\medef.h"
#include "m:\rid\mxdef.h"
#include "m:\rid\stdef.h"
static int cl_ptr (int ,char *,int ,char *,char **);
/* code -  cl_ass - assemble command line from vector */
int cl_ass(
char *str ,
int cnt ,
char *(*vec ))
{ 
#if Wnt
  char buf [mxLIN];
  char *cmd = buf;
  if( (cl_lin (cmd)) == 0)return ( 0 );
  while ( *cmd && *cmd != ' ') {++cmd ;}
  while ( *cmd && *cmd == ' ') {++cmd ;}
  st_cop (cmd, str);
  return 1;
#else 
  int nth = 0;
  ++vec, --cnt;
  *str = 0;
  while ( cnt--) {
    if( ((st_len (*vec) + st_len (str)) + 2) >= mxLIN)return 0;
    if ( nth++) {st_app (" ", str) ;}
    st_app (*vec++, str);
  } 
  return 1;
#endif 
} 
/* code -  cl_vec - convert to string vector */
int cl_vec(
char *str ,
int *cnt ,
char *(*(*ptr )))
{ int nth = 0;
  char *(*vec )= *ptr;
  char buf [256];
  if ( vec == NULL) {
    vec = me_alc (33 *  sizeof(char *));
    *ptr = vec; }
  *cnt = 0;
  *vec = st_dup ("");
  for(;;)  {
    *vec = NULL, *cnt = nth;
    if( nth == 32)return 0;
    if( (cl_nth (nth++, str, buf)) == 0)return 1;
    *vec++ = st_dup (buf);
  } 
  return 1;
} 
/* code -  cl_arg - get nth argument from string */
int cl_arg(
int nth ,
char *str ,
char *dst ,
int sep ,
char *bal )
{ char *src ;
  int len ;
  if( (len = cl_ptr(nth,str,sep,bal,&src)) == 0)return ( 0 );
  *(char *)me_mov (src,dst,len) = 0;
  return 1;
} 
/* code -  cl_ptr - get pointer to nth argument */
static int cl_ptr(
int nth ,
char *str ,
int sep ,
char *bal ,
char *(*res ))
{ if( nth < 0)return 0;
  if ( sep == 0) {sep = ' ' ;}
  if ( bal == 0) {bal = "\"\'" ;}
  for(;;)  {
    while ( ct_spc (*str)) {++str ;}
    if( *str == 0)break;
    *res = str;
    while ( *str != 0) {
      bal = "\"\'";
      while ( *bal != 0) {
        if ( *str == *bal++) {
           str = st_bal (str);break; } }
      if( *str == ' ' || !*str)break;
      ++str;
    } 
    if( nth-- == 0)break;
    if ( *str != 0) {++str ;}
  } 
  if( nth >= 0)return 0;
  return ( str-*res);
} 
